src/main.rs

```diff
@@ -1,11 +1,110 @@
-    }
+    tracing_subscriber::fmt::init();
+
+    println!("Launching trin");
+
+    let trin_config = TrinConfig::new();
+    trin_config.display_config();
+
+    let infura_project_id = match env::var("TRIN_INFURA_PROJECT_ID") {
+        Ok(val) => val,
+        Err(_) => panic!(
+            "Must supply Infura key as environment variable, like:\n\
+            TRIN_INFURA_PROJECT_ID=\"your-key-here\" trin"
+        ),
+    };
+
+    let bootnode_enrs = trin_config
+        .bootnodes
+        .iter()
+        .map(|nodestr| nodestr.parse().unwrap())
+        .collect();
+
+    let portalnet_config = PortalnetConfig {
+        external_addr: trin_config.external_addr,
+        private_key: trin_config.private_key.clone(),
+        listen_port: trin_config.discovery_port,
+        bootnode_enrs,
+        ..Default::default()
+    };
+
+    // Initialize state sub-network, if selected
+    let (state_tx, state_handler) = if trin_config.networks.iter().any(|val| val == "state") {
+        let (state_tx, state_rx) = mpsc::unbounded_channel::<StateNetworkEndpoint>();
+        let state_handler = match trin_state::initialize(state_rx) {
+            Ok(val) => val,
+            Err(msg) => panic!("Error while initializing state network: {:?}", msg),
+        };
+        (Some(state_tx), Some(state_handler))
+    } else {
+        (None, None)
+    };
+
+    // Initialize chain history sub-network, if selected
+    let (history_tx, history_handler) = if trin_config.networks.iter().any(|val| val == "history") {
+        let (history_tx, history_rx) = mpsc::unbounded_channel::<HistoryNetworkEndpoint>();
+        let history_handler = match trin_history::initialize(history_rx) {
+            Ok(val) => val,
+            Err(msg) => panic!("Error while initializing chain history network: {:?}", msg),
+        };
+        (Some(history_tx), Some(history_handler))
+    } else {
+        (None, None)
+    };
+
+    // Initialize jsonrpc server
+    let (jsonrpc_tx, jsonrpc_rx) = mpsc::unbounded_channel::<PortalEndpoint>();
+    let web3_server_task = tokio::task::spawn_blocking(|| {
+        launch_jsonrpc_server(trin_config, infura_project_id, jsonrpc_tx);
+    });
+
+    tokio::spawn(async move {
+        info!(
+            "About to spawn portal p2p with boot nodes: {:?}",
+            portalnet_config.bootnode_enrs
+        );
+
+        let (mut p2p, events) = PortalnetProtocol::new(portalnet_config).await.unwrap();
+
+        let rpc_handler = JsonRpcHandler {
+            discovery: p2p.discovery.clone(),
+            jsonrpc_rx,
+            state_tx,
+            history_tx,
+        };
+
+        tokio::spawn(events.process_discv5_requests());
+        tokio::spawn(rpc_handler.process_jsonrpc_requests());
+        if let Some(handler) = state_handler {
+            tokio::spawn(handler.handle_client_queries());
+        }
+        if let Some(handler) = history_handler {
+            tokio::spawn(handler.handle_client_queries());
+        }
```

Comment:
If you feel like pursuing it, the direction I'd take is something like building a vector of subnet handlers and looping through them to spawn. 👌🏻 if you want to merge without trying it though.